pinot-common/src/main/java/org/apache/pinot/common/utils/helix/HelixHelper.java

```diff
@@ -495,4 +499,104 @@ public IdealState apply(@Nullable IdealState idealState) {
+        LOGGER.info("Updating instance tags {} for instance: {}", instanceTags, instanceId);
+        updateInstanceConfig(helixManager, instanceConfig);
+      }
+    }
+  }
+
+  /**
+   * Returns the instance config for a specific instance ID
+   * @param helixManager the Helix manager
+   * @param instanceId the unique ID for instance
+   * @return An InstanceConfig that we can update
+   */
+  public static InstanceConfig getInstanceConfig(HelixManager helixManager, String instanceId) {
+    HelixAdmin admin = helixManager.getClusterManagmentTool();
+    String clusterName = helixManager.getClusterName();
+    return admin.getInstanceConfig(clusterName, instanceId);
+  }
+
+  /**
+   * Update instance config into Helix properly
+   * @param helixManager the HelixManager for access
+   * @param instanceConfig the updated Helix Config
+   */
+  public static void updateInstanceConfig(HelixManager helixManager, InstanceConfig instanceConfig) {
+    // NOTE: Use HelixDataAccessor.setProperty() instead of HelixAdmin.setInstanceConfig() because the latter explicitly
+    // forbids instance host/port modification
+    HelixDataAccessor helixDataAccessor = helixManager.getHelixDataAccessor();
+    Preconditions.checkState(
+        helixDataAccessor.setProperty(helixDataAccessor.keyBuilder().instanceConfig(instanceConfig.getId()), instanceConfig),
+        "Failed to update instance config");
+  }
+
+  /**
+   * Update Helix Host and Name if the values are reasonable.
+   * @param helixManager the Helix Manager to get control
+   * @param instanceId the Helix instance Id
+   * @param hostName the Host name
+   * @param hostPort the Host port
+   * @return true if it is updated
+   */
+  public static boolean updateHostNamePort(HelixManager helixManager, String instanceId, String hostName, int hostPort) {
+    InstanceConfig instanceConfig = getInstanceConfig(helixManager, instanceId);
+    if (Strings.isNullOrEmpty(hostName)) {
+      LOGGER.info("host is empty, skip updating helix host.");
+      return false;
+    }
+    boolean updated = false;
+    // Update host and port if needed
+    if (!String.valueOf(hostPort).equals(instanceConfig.getPort())) {
+      instanceConfig.setPort(String.valueOf(hostPort));
+      updated = true;
+    }
+    if (!hostName.equals(instanceConfig.getHostName())) {
+      instanceConfig.setHostName(hostName);
+      updated = true;
+    }
+    updateInstanceConfig(helixManager, instanceConfig);
+    return updated;
+  }
+  /**
+   * Update Helix Host and Name if the values are reasonable.
+   * @param helixManager the Helix Manager to get control
+   * @param instanceId the Helix instance Id
+   * @param hostName the Host name
+   * @param hostPort the Host port
+   * @return true if it is updated
+   */
+  public static boolean updateHostNamePort(HelixManager helixManager, String instanceId, String hostName, String hostPort) {
+    if (Strings.isNullOrEmpty(hostPort)) {
```

Comment:
Same here. Also why do we log error here, while log info in the other method?

Reply:
Right, all changed to `log.warn` which makes more sense